### ISSUE TITLE
bug: Deep extending `gitbrowse.remote_patterns` overwrites defaults

### ISSUE THREAD

**Issue by futile** [folke/snacks.nvim#164]
### Did you check docs and existing issues?

- [X] I have read all the snacks.nvim docs
- [X] I have updated the plugin to the latest version before submitting this issue
- [X] I have searched the existing issues of snacks.nvim
- [X] I have searched the existing issues of plugins related to this issue

### Neovim version (nvim -v)

NVIM v0.10.2

### Operating system/version

Darwin Kernel Version 24.1.0 arm64

### Describe the bug

Setting `opts.gitbrowse.remote_patterns` seems to cause the defaults to be overwritten, instead of adding to them.

Using this config:
```lua
    {
      "folke/snacks.nvim",
      priority = 1000,
      lazy = false,
      keys = {
        { "<leader>gB", function() Snacks.gitbrowse() end, desc = "Git Browse", },
      },
      opts = function(_, opts)
        opts.gitbrowse = vim.tbl_deep_extend("error", opts.gitbrowse or {}, {
          remote_patterns = {
            { "^ssh://git@some%.custom%.host/(.*)/(.*)%.git$",  "https://some.custom.host/projects/%1/repos/%2" },
          },
          url_patterns = {
            ["some%.custom%.host"] = {
              file = "/browse/{file}?at={branch}#{line}",
            },
          },
        })
      end,
    },
```

Causes git repos for `some.custom.host` to work, but all others seem to break now, for example GitHub:
```
   Info  15:07:14 notify.info Git Browse Opening [origin](https://git@github.com:futile/nixos-config.git/blob/main/dotfiles/nvim/lua/plugins/base.lua#L30)
```

Because the url pattern still seems to be correct, I assume that it's `remote_patterns` that broke.

I think it might be this `"force"` that causes the default `remote_patterns` to be replaced instead of extended:
https://github.com/folke/snacks.nvim/blob/985be4a759f6fe83e569679da431eeb7d2db5286/lua/snacks/init.lua#L76

### Steps To Reproduce

1. Use a config that sets or extends `opts.gitbrowse.remote_patterns`.
2. Use `Snacks.gitbrowse()` in a repo that should be covered by the defaults, but not by the used config.
3. Instead of properly opening using the defaults, only a log message is shown, with a wrong url.

### Expected Behavior

The defaults to still work when adding custom `remote_patterns`.

### Repro

```lua
vim.env.LAZY_STDPATH = ".repro"
load(vim.fn.system("curl -s https://raw.githubusercontent.com/folke/lazy.nvim/main/bootstrap.lua"))()

require("lazy.minit").repro({
  spec = {
    { "folke/snacks.nvim", 
opts = function(_, opts)
        opts.gitbrowse = vim.tbl_deep_extend("error", opts.gitbrowse or {}, {
          remote_patterns = {
            { "^ssh://git@some%.custom%.host/(.*)/(.*)%.git$",  "https://some.custom.host/projects/%1/repos/%2" },
          },
          url_patterns = {
            ["some%.custom%.host"] = {
              file = "/browse/{file}?at={branch}#{line}",
            },
          },
        })
      end,
 },
    -- add any other plugins here
  },
})
```

**Comment by dpetka2001** [comment]
`remote_patterns` is a list, so you should use `vim.list_extend` instead on `remote_patterns` explicitly.

**Comment by futile** [comment]
> `remote_patterns` is a list, so you should use `vim.list_extend` instead on `remote_patterns` explicitly.

Thanks, that's helpful! I guess then an open question could be whether this kind of extension should be the default (i.e., when using `opts = {}` instead of `opts = function(_, opts) ...`), because it seems more intuitive. Or maybe an example config could also help here.

**Comment by dpetka2001** [comment]
I don't think it should be the default because it depends on the object you want to operate. `url_patterns` for example is a `key/value` dictionary-like table. Users should use whichever is correct based on the object they want to operate. You could do
```lua
{
  "folke/snacks.nvim",
  opts = function(_, opts)
    opts.gitbrowse.remote_patterns = vim.list_extend(opts.gitbrowse.remote_patterns, {
      { "your_custom_pattern", "custom_replace" },
    })
  end,
}
```
Or if you use `lazy.nvim` as your package manager you can do the following
```lua
{
  "folke/snacks.nvim",
  opts_extend = { "gitbrowse.remote_patterns" },
  opts = {
    remote_patterns = {
      { "custom_pattern", "custom_replace" },
    },
  },
}
```
The option `opts_extend` of `lazy.nvim` the package manager allows users to define lists as regular tables and `lazy.nvim` will extend the corresponding list with your custom settings instead of overwriting as it would happen normally.

**Comment by folke** [comment]
That doesn't work like that. The extend can only work for config options that have already been set in a lazy spec.
But those options are not set anywhere and are just the defaults in snacks.
Just copy the patterns in your own config.

**Comment by futile** [comment]
> That doesn't work like that. The extend can only work for config options that have already been set in a lazy spec.
> But those options are not set anywhere and are just the defaults in snacks.

Yeah, that's what it looks like to me as well (still useful to know about `opts_extend`).

> Just copy the patterns in your own config.

Would it make sense to export `defaults` from the `gitbrowse` module so I could reference that? Otherwise I'll just copy the patterns, but that way I won't automatically get new patterns that are added to the upstream here.

**Comment by folke** [comment]
@futile you can now change the config after it has been resolved.
See https://github.com/folke/snacks.nvim/tree/main?tab=readme-ov-file#%EF%B8%8F-configuration

**Comment by futile** [comment]
Wow, awesome, thank you so much!!

**Comment by konradmalik** [comment]
FYI: this also solved my issue/request #34
